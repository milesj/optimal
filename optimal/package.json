{
  "name": "optimal",
  "type": "module",
  "version": "5.0.0-alpha.2",
  "description": "A system for building and validating defined object structures with schemas.",
  "main": "./lib/node/index.js",
  "browser": "./lib/browser/index.js",
  "module": "./esm/index.js",
  "types": "./dts/index.d.ts",
  "keywords": [
    "opts",
    "options",
    "schema",
    "predicate",
    "validator",
    "config",
    "defaults",
    "object",
    "params",
    "struct"
  ],
  "repository": {
    "type": "git",
    "url": "git@github.com:milesj/optimal.git",
    "directory": "optimal"
  },
  "author": {
    "name": "Miles Johnson",
    "url": "http://milesj.me"
  },
  "license": "MIT",
  "files": [
    "dts/**/*.d.ts",
    "esm/**/*.{js,map}",
    "lib/**/*.{js,map}",
    "src/**/*.{ts,tsx,json}"
  ],
  "engines": {
    "node": ">=12.17.0",
    "npm": ">=6.13.0"
  },
  "funding": {
    "type": "ko-fi",
    "url": "https://ko-fi.com/milesjohnson"
  },
  "packemon": {
    "bundle": true,
    "format": [
      "lib",
      "esm"
    ],
    "platform": [
      "browser",
      "node"
    ]
  }
}
