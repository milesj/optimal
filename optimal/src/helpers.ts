import {
	Constructor,
	DefaultValue,
	DefaultValueInitializer,
	Schema,
	SchemaValidateOptions,
	UnknownObject,
} from './types';
import { ValidationError } from './ValidationError';

export function isObject(value: unknown): value is object {
	return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
	return (
		isObject(value) &&
		typeof (value as UnknownObject).schema === 'function' &&
		typeof (value as UnknownObject).state === 'function' &&
		typeof (value as UnknownObject).type === 'function' &&
		typeof (value as UnknownObject).validate === 'function'
	);
}

export function isValidDate(value: Date): value is Date {
	return (
		Object.prototype.toString.call(value) === '[object Date]' &&
		!Number.isNaN(value.getTime()) &&
		value.toString() !== 'Invalid Date'
	);
}

export function isValidNumber(value: unknown): value is number {
	return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
	return typeof value === 'string' && value !== '';
}

export function createArray(value: unknown): unknown[] {
	if (value === undefined) {
		return [];
	}

	return Array.isArray(value) ? [...(value as unknown[])] : [value];
}

export function createDate(value: unknown): Date {
	if (value instanceof Date) {
		return value;
	}

	if (value === undefined || value === null) {
		return new Date();
	}

	return new Date(value as number);
}

export function createObject<T = UnknownObject>(value: unknown) {
	return (isObject(value) ? { ...value } : {}) as unknown as T;
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(object: unknown, contract: Constructor<T>): object is T {
	if (!object || typeof object !== 'object') {
		return false;
	}

	if (object instanceof contract) {
		return true;
	}

	let current = object;

	while (current) {
		if (current.constructor.name === 'Object') {
			return false;
		}

		if (
			current.constructor.name === contract.name ||
			(current instanceof Error && current.name === contract.name)
		) {
			return true;
		}

		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
		current = Object.getPrototypeOf(current);
	}

	return false;
}

export function invalid(
	condition: boolean,
	message: string,
	path: string = '',
	value: unknown = undefined,
) {
	if (condition) {
		return;
	}

	throw new ValidationError(message, path, value);
}

export function invariant(condition: boolean, message: string) {
	if (condition) {
		return;
	}

	throw new Error(message);
}

export function logUnknown(unknownFields: object, pathPrefix?: string) {
	const unknownKeys = Object.keys(unknownFields);

	if (unknownKeys.length > 0) {
		throw new Error(
			`${pathPrefix ? `Unknown "${pathPrefix}" fields` : 'Unknown fields'}: ${unknownKeys.join(
				', ',
			)}.`,
		);
	}
}

export function pathKey(path: string): string {
	if (path.endsWith(']')) {
		const index = path.lastIndexOf('[');

		return index > 0 ? path.slice(index) : path;
	}

	const index = path.lastIndexOf('.');

	return index > 0 ? path.slice(index + 1) : path;
}

export function prettyValue(value: unknown): string | null {
	switch (typeof value) {
		case 'string':
			return `"${value}"`;

		case 'number':
		case 'function':
			return String(value);

		case 'object': {
			if (value === null) {
				return `\`null\``;
			}

			if (value.constructor !== Object) {
				return value.constructor.name === 'Array' ? null : `\`${value.constructor.name}\``;
			}

			return null;
		}

		default:
			return `\`${value}\``;
	}
}

export function tryAndCollect(
	validator: () => boolean | void,
	validError: ValidationError,
	collectErrors?: boolean,
): boolean {
	let result = false;

	try {
		const value = validator();

		if (typeof value === 'boolean') {
			result = value;
		}
	} catch (error: unknown) {
		if (error instanceof Error && collectErrors) {
			validError.addError(error);
		} else {
			throw error;
		}
	}

	return result;
}

export function extractDefaultValue<T>(
	defaultValue: DefaultValue<T>,
	path: string,
	{ currentObject, rootObject }: SchemaValidateOptions,
) {
	return typeof defaultValue === 'function'
		? (defaultValue as DefaultValueInitializer<T>)(path, currentObject!, rootObject!)
		: defaultValue;
}
