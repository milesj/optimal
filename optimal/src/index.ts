/**
 * @copyright   2021, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

import * as schemas from './schemas';

export type Schemas = typeof schemas;

export { schemas };
export * from './createPredicate';
export * from './createSchema';
export * from './criteria';
export * from './optimal';
export * from './OptimalError';
export * from './schemas/array';
export * from './schemas/blueprint';
export * from './schemas/bool';
export * from './schemas/custom';
export * from './schemas/date';
export * from './schemas/func';
export * from './schemas/instance';
export * from './schemas/lazy';
export * from './schemas/number';
export * from './schemas/object';
export * from './schemas/regex';
export * from './schemas/schema';
export * from './schemas/shape';
export * from './schemas/string';
export * from './schemas/tuple';
export * from './schemas/union';
export * from './types';
export * from './ValidationError';
