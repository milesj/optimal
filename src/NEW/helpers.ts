import { Constructor, UnknownObject, Schema } from './types';

export function createDate(value: unknown): Date {
  return value instanceof Date ? value : new Date(value as number);
}

/**
 * Native `instanceof` checks are problematic, as cross realm checks fail.
 * They will also fail when comparing against source and compiled files.
 * So emulate an `instanceof` check by comparing constructor names.
 */
export function instanceOf<T = unknown>(
  object: unknown,
  contract: Constructor<T> | Function,
): object is T {
  if (!object || typeof object !== 'object') {
    return false;
  }

  if (object instanceof contract) {
    return true;
  }

  let current = object;

  while (current) {
    if (current.constructor.name === 'Object') {
      return false;
    }

    if (
      current.constructor.name === contract.name ||
      (current instanceof Error && current.name === contract.name)
    ) {
      return true;
    }

    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    current = Object.getPrototypeOf(current);
  }

  return false;
}

export function invariant(condition: boolean, message: string, path: string = '') {
  if (__DEV__) {
    if (condition) {
      return;
    }

    // TODO no errorPrefix
    const prefix = path ? `Invalid field "${path}".` : '';

    throw new Error(`${prefix} ${message}`.trim());
  }
}

export function isObject(value: unknown): value is object {
  return !!value && typeof value === 'object' && !Array.isArray(value);
}

export function isSchema<T>(value: unknown): value is Schema<T> {
  return (
    isObject(value) &&
    typeof (value as UnknownObject).validate === 'function' &&
    typeof (value as UnknownObject).type === 'function'
  );
}

export function isValidDate(value: Date): value is Date {
  return (
    Object.prototype.toString.call(value) === '[object Date]' &&
    !Number.isNaN(value.getTime()) &&
    value.toString() !== 'Invalid Date'
  );
}

export function isValidNumber(value: unknown): value is number {
  return typeof value === 'number' && !Number.isNaN(value);
}

export function isValidString(value: unknown): value is string {
  return typeof value === 'string' && value !== '';
}

export function pathKey(path: string): string {
  const index = path.lastIndexOf('.');

  return index > 0 ? path.slice(index + 1) : path;
}
