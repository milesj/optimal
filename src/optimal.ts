import Schema from './Schema';
import isObject from './isObject';
import { Blueprint, OptimalOptions } from './types';

export default function optimal<
  Struct extends object,
  Construct extends object = { [K in keyof Struct]?: unknown }
>(struct: Construct, blueprint: Blueprint<Struct>, options: OptimalOptions = {}): Required<Struct> {
  if (__DEV__) {
    if (!isObject(options)) {
      throw new TypeError('Optimal options must be a plain object.');
    }
  }

  const schema = new Schema(blueprint);

  if (options.name) {
    schema.setName(options.name);
  }

  if (options.file) {
    schema.setFile(options.file);
  }

  if (options.unknown) {
    schema.allowUnknown();
  }

  return schema.build(struct, options.prefix);
}
