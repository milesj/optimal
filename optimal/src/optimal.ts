import { isObject } from './helpers';
import { OptimalError } from './OptimalError';
import { shape } from './schemas/shape';
import { Blueprint, DeepPartial, SchemaValidateOptions, UnknownObject } from './types';
import { ValidationError } from './ValidationError';

export interface OptimalOptions {
	file?: string;
	name?: string;
	prefix?: string;
	unknown?: boolean;
}

export interface Optimal<T extends object> {
	configure: (options: OptimalOptions) => void;
	validate: (struct: DeepPartial<T>, options?: SchemaValidateOptions) => Required<T>;
}

export function optimal<Schemas extends object>(
	blueprint: Blueprint<Schemas>,
	baseOpts: OptimalOptions = {},
): Optimal<Schemas> {
	const options: OptimalOptions = {};
	let schema = shape(blueprint);

	function configure(nextOpts: OptimalOptions) {
		if (!isObject(nextOpts)) {
			throw new TypeError('Optimal options must be a plain object.');
		}

		Object.assign(options, nextOpts);
		schema = schema.exact(!options.unknown);
	}

	configure(baseOpts);

	return {
		configure,
		validate(struct, validateOptions) {
			const object = struct as UnknownObject;

			try {
				return schema.validate(struct, options.prefix ?? '', {
					...validateOptions,
					currentObject: object,
					rootObject: object,
				});
			} catch (error: unknown) {
				let invalid: OptimalError;
				let prefix = '';

				if (error instanceof OptimalError) {
					invalid = error;
				} else {
					invalid = new OptimalError();
					invalid.addError(error as ValidationError);
				}

				if (options.name) {
					invalid.schema = options.name;
					prefix = options.name;
				}

				if (options.file) {
					invalid.file = options.file;
					prefix = prefix ? `${prefix} (${options.file})` : options.file;
				}

				if (prefix) {
					invalid.message = `${prefix}: ${invalid.message}`;
				}

				throw invalid;
			}
		},
	};
}
