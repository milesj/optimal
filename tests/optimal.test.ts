import optimal, {
  array,
  bool,
  func,
  instance,
  number,
  object,
  regex,
  shape,
  string,
  union,
  ObjectOf,
} from '../src';
import { runInProd } from './helpers';

class Plugin {}

describe('Optimal', () => {
  type PrimitiveType = string | number | boolean;
  type ConditionType = string | Function | RegExp | RegExp[] | ObjectOf<RegExp>;

  // This blueprint is based on Webpack's configuration: https://webpack.js.org/configuration/
  // Webpack provides a pretty robust example of how to use this library.
  const primitive = union<PrimitiveType>([string(), number(), bool()], false);

  const condition = union<ConditionType>(
    [string(), func(), regex(), array(regex()), object(regex())],
    '',
  );

  const rule = shape({
    enforce: string('post').oneOf<'pre' | 'post'>(['pre', 'post']),
    exclude: condition,
    include: condition,
    issuer: condition,
    parser: object(bool()),
    resource: condition,
    use: array(
      union<string | object>(
        [
          string(),
          shape({
            loader: string(),
            options: object(primitive),
          }),
        ],
        [],
      ),
    ),
  });

  type EntryType = string | string[] | ObjectOf<string | string[]> | Function;
  type CrossOriginType = 'anonymous' | 'use-credentials';
  type HashType = 'md5' | 'sha256' | 'sha512';
  type NoParseType = RegExp | RegExp[] | Function;
  type TargetType =
    | 'async-node'
    | 'electron-main'
    | 'electron-renderer'
    | 'node'
    | 'node-webkit'
    | 'web'
    | 'webworker';
  type NodeType = 'mock' | 'empty';

  const blueprint = {
    context: string(process.cwd()),
    entry: union<EntryType>(
      [string(), array(string()), object(union([string(), array(string())], '')), func()],
      [],
    ).nullable(),
    output: shape({
      chunkFilename: string('[id].js'),
      chunkLoadTimeout: number(120000),
      crossOriginLoading: union<false | CrossOriginType>(
        [
          bool(false).only(),
          string('anonymous').oneOf<CrossOriginType>(['anonymous', 'use-credentials']),
        ],
        false,
      ),
      filename: string('bundle.js'),
      hashFunction: string('md5').oneOf<HashType>(['md5', 'sha256', 'sha512']),
      path: string(),
      publicPath: string(),
    }),
    module: shape({
      noParse: union<NoParseType | null>([regex(), array(regex()), func()], null).nullable(),
      rules: array(rule),
    }),
    resolve: shape({
      alias: object(string()),
      extensions: array(string()),
      plugins: array(instance(Plugin)),
      resolveLoader: object(array(string())),
    }),
    plugins: array(instance(Plugin)),
    target: string('web').oneOf<TargetType>([
      'async-node',
      'electron-main',
      'electron-renderer',
      'node',
      'node-webkit',
      'web',
      'webworker',
    ]),
    watch: bool(false),
    node: object(
      union<boolean | NodeType>(
        [bool(), string('mock').oneOf<NodeType>(['mock', 'empty'])],
        false,
      ),
    ),
  };

  it('errors if a non-object is passed', () => {
    expect(() => {
      optimal([], {});
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      optimal(123, {});
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      // @ts-ignore
      optimal('foo', {});
    }).toThrowErrorMatchingSnapshot();

    expect(() => {
      optimal(() => {}, {});
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if a non-object is passed as a blueprint', () => {
    expect(() => {
      // @ts-ignore
      optimal({}, 123);
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if a non-predicate is passed within the blueprint', () => {
    expect(() => {
      optimal(
        {},
        {
          // @ts-ignore
          foo: 123,
        },
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('errors if a non-object config is passed', () => {
    expect(() => {
      // @ts-ignore
      optimal({}, blueprint, 123);
    }).toThrowErrorMatchingSnapshot();
  });

  it('sets object keys as class properties', () => {
    const options = optimal<{
      foo: number;
      bar: boolean;
      baz: string;
    }>(
      {
        foo: 123,
        bar: true,
      },
      {
        foo: number(0),
        bar: bool(true),
        baz: string(),
      },
    );

    expect(options.foo).toBe(123);
    expect(options.bar).toBe(true);
    expect(options.baz).toBe('');
    expect(options).toEqual({
      foo: 123,
      bar: true,
      baz: '',
    });
  });

  it('sets default values', () => {
    const options = optimal({}, blueprint);

    expect(options).toEqual({
      context: process.cwd(),
      entry: [],
      output: {
        chunkFilename: '[id].js',
        chunkLoadTimeout: 120000,
        crossOriginLoading: false,
        filename: 'bundle.js',
        hashFunction: 'md5',
        path: '',
        publicPath: '',
      },
      module: {
        noParse: null,
        rules: [],
      },
      resolve: {
        alias: {},
        extensions: [],
        plugins: [],
        resolveLoader: {},
      },
      plugins: [],
      target: 'web',
      watch: false,
      node: {},
    });
  });

  it('runs checks for root level values', () => {
    expect(() => {
      optimal(
        {
          entry: 123,
        },
        blueprint,
      );
    }).toThrow(
      'Invalid field "entry". Type must be one of: string, array<string>, object<string | array<string>>, function',
    );
  });

  it('runs checks for nested level values', () => {
    expect(() => {
      optimal(
        {
          output: {
            crossOriginLoading: 'not-anonymous',
          },
        },
        blueprint,
      );
    }).toThrowErrorMatchingSnapshot();
  });

  it('includes a custom `name` in the error message', () => {
    expect(() => {
      optimal(
        {
          entry: 123,
        },
        blueprint,
        {
          name: 'FooBar',
        },
      );
    }).toThrow(
      'Invalid FooBar field "entry". Type must be one of: string, array<string>, object<string | array<string>>, function',
    );
  });

  describe('production', () => {
    it(
      'sets and returns correct propertes',
      runInProd(() => {
        const options = optimal(
          {
            entry: ['foo.js'],
            output: {
              hashFunction: 'sha256',
            },
            module: {
              noParse: /foo/u,
            },
            // Invalid, should not error
            target: 'unknown',
          },
          blueprint,
        );

        expect(options).toEqual({
          context: process.cwd(),
          entry: ['foo.js'],
          output: {
            chunkFilename: '[id].js',
            chunkLoadTimeout: 120000,
            crossOriginLoading: false,
            filename: 'bundle.js',
            hashFunction: 'sha256',
            path: '',
            publicPath: '',
          },
          module: {
            noParse: /foo/u,
            rules: [],
          },
          resolve: {
            alias: {},
            extensions: [],
            plugins: [],
            resolveLoader: {},
          },
          plugins: [],
          target: 'unknown',
          watch: false,
          node: {},
        });
      }),
    );
  });

  describe('unknown fields', () => {
    it('errors for unknown fields', () => {
      expect(() => {
        optimal(
          {
            foo: 123,
            bar: 456,
          },
          blueprint,
        );
      }).toThrowErrorMatchingSnapshot();
    });

    it('doesnt error for unknown fields if `unknown` is true', () => {
      expect(() => {
        optimal(
          {
            foo: 123,
            bar: 456,
          },
          blueprint,
          {
            unknown: true,
          },
        );
      }).not.toThrow();
    });

    it('sets unknown fields', () => {
      expect(
        optimal(
          {
            foo: 123,
            bar: 456,
          },
          blueprint,
          {
            unknown: true,
          },
        ),
      ).toEqual(
        expect.objectContaining({
          foo: 123,
          bar: 456,
        }),
      );
    });
  });

  describe('logical operators', () => {
    it('handles AND', () => {
      const and = {
        foo: string('a').and('bar', 'baz'),
        bar: string('b').and('foo', 'baz'),
        baz: string('c').and('foo', 'bar'),
      };

      // Dont error if all are undefined
      expect(() => {
        optimal(
          {},
          {
            foo: string('a').and('bar', 'baz'),
            bar: string('b').and('foo', 'baz'),
            baz: string('c').and('foo', 'bar'),
          },
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            foo: 'a',
          },
          and,
        );
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        optimal(
          {
            foo: 'a',
            bar: 'b',
          },
          and,
        );
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        optimal(
          {
            foo: 'a',
            baz: 'c',
          },
          and,
        );
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        optimal(
          {
            foo: 'a',
            bar: 'b',
            baz: 'c',
          },
          and,
        );
      }).not.toThrow();
    });

    it('handles OR', () => {
      const or = {
        foo: string('a').or('bar', 'baz'),
        bar: string('b').or('foo', 'baz'),
        baz: string('c').or('foo', 'bar'),
      };

      expect(() => {
        optimal({}, or);
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        optimal(
          {
            foo: 'a',
          },
          or,
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            bar: 'b',
          },
          or,
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            baz: 'c',
          },
          or,
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            foo: 'a',
            bar: 'b',
            baz: 'c',
          },
          or,
        );
      }).not.toThrow();
    });

    it('handles XOR', () => {
      const xor = {
        foo: string('a').xor('bar', 'baz'),
        bar: string('b').xor('foo', 'baz'),
        baz: string('c').xor('foo', 'bar'),
      };

      expect(() => {
        optimal({}, xor);
      }).toThrowErrorMatchingSnapshot();

      expect(() => {
        optimal(
          {
            foo: 'a',
          },
          xor,
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            bar: 'b',
          },
          xor,
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            baz: 'c',
          },
          xor,
        );
      }).not.toThrow();

      expect(() => {
        optimal(
          {
            foo: 'a',
            bar: 'b',
            baz: 'c',
          },
          xor,
        );
      }).toThrowErrorMatchingSnapshot();
    });
  });
});
