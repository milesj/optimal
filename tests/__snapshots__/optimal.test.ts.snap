// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Optimal errors if a non-object config is passed 1`] = `"Optimal options must be a plain object."`;

exports[`Optimal errors if a non-object is passed 1`] = `"Schema requires a plain object, found array."`;

exports[`Optimal errors if a non-object is passed 2`] = `"Schema requires a plain object, found number."`;

exports[`Optimal errors if a non-object is passed 3`] = `"Schema requires a plain object, found string."`;

exports[`Optimal errors if a non-object is passed 4`] = `"Schema requires a plain object, found function."`;

exports[`Optimal errors if a non-object is passed as a blueprint 1`] = `"A schema blueprint is required."`;

exports[`Optimal errors if a non-predicate is passed within the blueprint 1`] = `"Unknown blueprint for \\"foo\\". Must be a predicate."`;

exports[`Optimal logical operators handles AND 1`] = `"All of these fields must be defined: foo, bar, baz"`;

exports[`Optimal logical operators handles AND 2`] = `"All of these fields must be defined: foo, bar, baz"`;

exports[`Optimal logical operators handles AND 3`] = `"All of these fields must be defined: foo, bar, baz"`;

exports[`Optimal logical operators handles OR 1`] = `"At least one of these fields must be defined: foo, bar, baz"`;

exports[`Optimal logical operators handles XOR 1`] = `"Only one of these fields may be defined: foo, bar, baz"`;

exports[`Optimal logical operators handles XOR 2`] = `"Only one of these fields may be defined: foo, bar, baz"`;

exports[`Optimal runs checks for nested level values 1`] = `
"Invalid field \\"output.crossOriginLoading\\". Type must be one of: boolean, string. Received string with the following invalidations:
 - String must be one of: anonymous, use-credentials"
`;

exports[`Optimal unknown fields errors for unknown fields 1`] = `"Unknown fields: foo, bar."`;
